### PR TITLE
Better parser using megaparsec

### LINKED ISSUE #29: src/Language/Docker/PrettyPrint.hs:91:25: error: Not in scope: 'fail'
Trying to build  4.0.1 here with ghc 8.0.1, and seeing this issue:

```
./setup build
Building language-docker-4.0.1...
Preprocessing library language-docker-4.0.1...
[ 1 of 11] Compiling Paths_language_docker ( dist/build/autogen/Paths_language_docker.hs, dist/build/Paths_language_docker.o )
[ 2 of 11] Compiling Language.Docker.Syntax ( src/Language/Docker/Syntax.hs, dist/build/Language/Docker/Syntax.o )
[ 3 of 11] Compiling Language.Docker.Syntax.Lift ( src/Language/Docker/Syntax/Lift.hs, dist/build/Language/Docker/Syntax/Lift.o )
[ 4 of 11] Compiling Language.Docker.PrettyPrint ( src/Language/Docker/PrettyPrint.hs, dist/build/Language/Docker/PrettyPrint.o )

src/Language/Docker/PrettyPrint.hs:91:25: error:
    Not in scope: ‘fail’
    Perhaps you want to add ‘fail’ to the import list in the import of
    ‘Prelude’ (src/Language/Docker/PrettyPrint.hs:(15,1)-(17,18)).
[ 5 of 11] Compiling Language.Docker.Normalize ( src/Language/Docker/Normalize.hs, dist/build/Language/Docker/Normalize.o )
[ 6 of 11] Compiling Language.Docker.Lexer ( src/Language/Docker/Lexer.hs, dist/build/Language/Docker/Lexer.o )
[ 7 of 11] Compiling Language.Docker.Parser ( src/Language/Docker/Parser.hs, dist/build/Language/Docker/Parser.o )
[ 8 of 11] Compiling Language.Docker.EDSL.Types ( src/Language/Docker/EDSL/Types.hs, dist/build/Language/Docker/EDSL/Types.o )
```

### LINKED ISSUE #32: Parse RUN agruments do not respect control operators
## Intro
`bash` manpage:
```
       control operator
              A token that performs a control function.  It is one of the following symbols:
              || & && ; ;; ( ) | |& <newline>
```

Control operators are important for decisions which are happening later in `hadolint` https://github.com/hadolint/hadolint/blob/master/src/Hadolint/Rules.hs#L173

This issue has been found through https://github.com/hadolint/hadolint/issues/196

## Current state
```dockerfile
RUN set -ex; apk add --no-cache curl=7.59.0-r0
```

is parsed
```
$ stack repl
> parseString "RUN set -ex; apk add --no-cache ca-certificates=20171114-r0"
Right [InstructionPos {instruction = Run (Arguments ["set","-ex;","apk","add","--no-cache","ca-certificates=20171114-r0"]), sourcename = "<string>", lineNumber = 1}]
```
where `;` is parsed as part of `-ex`.

I believe it is done here
https://github.com/hadolint/language-docker/blob/20c45d269514e45d90cb2eb06c0deb20cdcd6454/src/Language/Docker/Parser.hs#L352

## Desired state

Split `RUN` arguments using white space and selected control operators.

```
$ stack repl
> parseString "RUN set -ex; apk add --no-cache ca-certificates=20171114-r0"
Right [InstructionPos {instruction = Run (Arguments ["set","-ex",";","apk","add","--no-cache","ca-certificates=20171114-r0"]), sourcename = "<string>", lineNumber = 1}]
```

### PR DESCRIPTION
**Pull request by lorenzo**
I had quite a few improvements that I wanted to add to the parser, so I decided to bundle them all together, this PR includes:

* A faster parser thanks to the use of `megaparsec` instead of `parsec`
* Use `Text` instead of `String` were possible, this also improved performance
* IO operations are done though the `bytestring` package instead of the Prelude, and always assumes UTF-8. This will eliminate the errors caused by locale mismatches in the env.
* Changed the representation of `Arguments` to have a type variable. This will help `hadolint` parse the instructions in the dockerfile using shellcheck without any crazy strings juggling.
* Made it compile again in older GHC 8.0.2

fixes #29 
fixes #32
refs https://github.com/hadolint/hadolint/issues/203
refs https://github.com/hadolint/hadolint/issues/198

### CODE REVIEW

File: examples/templating-7.10.dockerfile [hunk 1]
```diff
@@ -4,4 +4,4 @@ RUN cabal update
 ADD mypackage.cabal /app/mypackage.cabal
 RUN cabal install --only-dep -j
 ADD . /app/
-RUN cabal build
+RUN cabal build
```
Comment (zemanlx) [inline examples/templating-7.10.dockerfile]: Is there a reason for removing *newlines* from example files?

Comment (lorenzo) [reply examples/templating-7.10.dockerfile]: I think that was done by running the examples themselves. Looks like the new script is not adding a newline. Do you think we should add it by default?

File: src/Language/Docker/PrettyPrint.hs [hunk 2]
```diff
@@ -69,14 +74,15 @@ prettyPrintPairs ps = hsep $ fmap prettyPrintPair ps
...
+prettyPrintArguments :: Arguments Text -> Doc ann
+prettyPrintArguments (ArgumentsList as) = prettyPrintJSON (Text.words as)
+prettyPrintArguments (ArgumentsText as) = hsep (fmap helper (Text.words as))
   where
     helper "&&" = "\\\n &&"
     helper a = pretty a
 
-prettyPrintJSON :: Arguments -> Doc ann
-prettyPrintJSON (Arguments as) = list (fmap doubleQoute as)
+prettyPrintJSON :: [Text] -> Doc ann
```
[the remark this answers is not in the document]
Comment (lorenzo) [reply src/Language/Docker/PrettyPrint.hs]: Yeah, it is still used here https://github.com/hadolint/language-docker/pull/34/files#diff-76730fcc0b0a28453e6ef933dbc53edbR77

**Comment by lorenzo** [comment]
@zemanlx anything I can do to help you understand more?

File: examples/templating-7.10.dockerfile (hunk 1 above)
Comment (lorenzo) [reply examples/templating-7.10.dockerfile]: I'll actually go ahead and do that, I think preserving the previous behaviour (automatically adding a new line) is important

**Comment by zemanlx** [comment]
Not really, it's just huge space between my level and your level of Haskell knowledge/experience.

**Pull request merged**